h2/src/main/org/h2/expression/IntervalOperation.java

```diff
@@ -0,0 +1,320 @@
+            case Value.DATE:
+            case Value.TIMESTAMP:
+            case Value.TIMESTAMP_TZ:
+                if (DataType.isYearMonthIntervalType(rType)) {
+                    long m = DateTimeUtils.intervalToAbsolute((ValueInterval) r).longValue();
+                    if (opType == IntervalOpType.DATETIME_MINUS_INTERVAL) {
+                        m = -m;
+                    }
+                    return DateTimeFunctions.dateadd("MONTH", m, l);
+                } else {
+                    BigInteger a2 = DateTimeUtils.intervalToAbsolute((ValueInterval) r);
+                    if (lType == Value.DATE) {
+                        BigInteger a1 = BigInteger
+                                .valueOf(DateTimeUtils.absoluteDayFromDateValue(((ValueDate) l).getDateValue()));
+                        a2 = a2.divide(BigInteger.valueOf(DateTimeUtils.NANOS_PER_DAY));
+                        BigInteger n = opType == IntervalOpType.DATETIME_PLUS_INTERVAL ? a1.add(a2) : a1.subtract(a2);
+                        return ValueDate.fromDateValue(DateTimeUtils.dateValueFromAbsoluteDay(n.longValue()));
+                    } else {
+                        long[] a = DateTimeUtils.dateAndTimeFromValue(l);
+                        long absoluteDay = DateTimeUtils.absoluteDayFromDateValue(a[0]);
+                        long timeNanos = a[1];
+                        BigInteger[] dr = a2.divideAndRemainder(BigInteger.valueOf(DateTimeUtils.NANOS_PER_DAY));
+                        if (opType == IntervalOpType.DATETIME_PLUS_INTERVAL) {
+                            absoluteDay += dr[0].longValue();
+                            timeNanos += dr[1].longValue();
+                        } else {
+                            absoluteDay -= dr[0].longValue();
+                            timeNanos -= dr[1].longValue();
+                        }
+                        if (timeNanos >= DateTimeUtils.NANOS_PER_DAY) {
+                            timeNanos -= DateTimeUtils.NANOS_PER_DAY;
+                            absoluteDay++;
+                        } else if (timeNanos < 0) {
+                            timeNanos += DateTimeUtils.NANOS_PER_DAY;
+                            absoluteDay--;
+                        }
+                        return DateTimeUtils.dateTimeToValue(l, DateTimeUtils.dateValueFromAbsoluteDay(absoluteDay),
+                                timeNanos, false);
+                    }
+                }
+            }
+            break;
+        case INTERVAL_MULTIPLY_NUMERIC:
+        case INTERVAL_DIVIDE_NUMERIC: {
+            BigDecimal a1 = new BigDecimal(DateTimeUtils.intervalToAbsolute((ValueInterval) l));
+            BigDecimal a2 = r.getBigDecimal();
+            return DateTimeUtils.intervalFromAbsolute(IntervalQualifier.valueOf(lType - Value.INTERVAL_YEAR),
+                    (opType == IntervalOpType.INTERVAL_MULTIPLY_NUMERIC ? a1.multiply(a2) : a1.divide(a2))
+                            .toBigInteger());
```

Comment:
overflow check?

Reply:
`BigDecimal` / `BigInteger` are used here.

(But I found one missing check in `intervalFromAbsolute()`.)